model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java

```diff
@@ -617,20 +626,11 @@ public int getUsersCount(RealmModel realm) {
 
     @Override
     public List<UserModel> getUsers(RealmModel realm, int firstResult, int maxResults, boolean includeServiceAccounts) {
-        String queryName = includeServiceAccounts ? "getAllUsersByRealm" : "getAllUsersByRealmExcludeServiceAccount" ;
+        Map<String, String> filters = new HashMap<>();
```

Comment:
We want Hibernate to provide values of the fillters as parameters, hence values should be `Object`s, not `String`s.

Reply:
Do you mean changing https://github.com/keycloak/keycloak/blob/5e779bda842a7c725e7c09eb0499930e538dd5ed/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L703 signature ?

Reply:
I believe we should.

The current changes in `UsersResources` are beyond the API contract already, since according to the JavaDoc documentation, the `attributes` can only contain `first`, `last`, `username`, and `email`, and for these the type would stay `String`.

So if we are to fix this issue, we need to change the API contract anyway. Hence my take is to change the signature to make it compile time rather than runtime issue, add a big fat note into migration guide that this change takes place and that the attributes can contain other attribute names than what has been specified in `org.keycloak.storage.user.UserQueryProvider.searchForUser` javadocs, and that the code needs to check value type as well.

@stianst Do you agree with this change? Could you please comment here?

Reply:
Continuing discussion below.